### ISSUE TITLE
Support for variable-length input arrays

### ISSUE THREAD

**Issue by msqr** [pjungwir/aggs_for_vecs#2]
My application has array columns that can vary in length. Mostly they are the same length, but they _can_ differ. When I query them, however, I know up front the _maximum_ length of all rows for the column. I was wondering how feasible it would be to add support for passing an array length into the aggregate functions, like

```
SELECT vec_to_sum(array_column, 8)
FROM foo
```

If the length is provided, then the aggregates would process each row and only handle that many elements, always returning an array of the given length (8 in my example). If no length is provided it would operate as it currently does, taking the length from the first input row.

**Comment by pjungwir** [comment]
Hmm, I don't love adding a parameter here for that, and then having to maintain two versions of each function. At the least I'd rather use a different name (e.g. `varvec_to_sum`) so that we can still add new parameters for something else someday. But before that, how about we write a helper function that expands an input array by padding it will NULLs so that it is some fixed length? So you could do `SELECT vec_to_sum(pad_vec(array_column, 8))`. It would be an error to give `pad_vec` an array longer than the passed in integer.

That won't be as fast as your proposal, but I'd like to try it out and see if it's fast enough. I don't think it will be very expensive.

I'm happy to do the work on that one. Perhaps this weekend if I get the chance.

**Comment by msqr** [comment]
Ah, the padding function sounds interesting, thanks! Given I expect most rows to have the same size, would it work out that the function could simply return the input array without having to make a copy? That sounds pretty great, if so.

**Comment by pjungwir** [comment]
I don't think that would cause any problems. Seems like a nice optimization!

**Comment by pjungwir** [comment]
Okay I added this function on a `pad_vec` branch.

Adding new functions is awkward, especially when collaborating, because they need to go into a new `aggs_for_vecs--some-new-version.sql` file, but you don't want to make that file in every change and then get merge conflicts. I made a `rel-1.3.0` branch that has a copy of it. I think I'll merge that into `master` and then we can rebase things off that (e.g. this function).
